src/Orleans/Messaging/Message.cs

```diff
@@ -446,7 +446,10 @@ private static object DeserializeBody(List<ArraySegment<byte>> bytes)
 
         public Message()
         {
-            headers = new Dictionary<Header, object>();
+            // average headers items count is 14 items, and while the Header enum contains 18 entries
+            // the closest prime number is 17; assuming that possibility of all 18 headers being at the same time is low enough to
+            // choose 17 in order to avoid allocations of two additional items on each call, and allocate 37 instead of 19 in rare cases
+            headers = new Dictionary<Header, object>(17);
```

Comment:
That would mean allocation of additional 20 slots (37 - 17) in underlying array, which will increase already pretty high memory traffic, as well as CPU usage (though not much: additional ~ 25 ms per 1 million iterations for 37 case compared to 17 on 3.1 GHz processor)